homeassistant/components/netatmo/light.py

```diff
@@ -32,61 +32,68 @@ async def async_setup_entry(
 ) -> None:
     """Set up the Netatmo camera light platform."""
-    data_handler = hass.data[DOMAIN][entry.entry_id][DATA_HANDLER]
-    data_class = data_handler.data.get(CAMERA_DATA_CLASS_NAME)
-
-    if not data_class or data_class.raw_data == {}:
-        raise PlatformNotReady
-
-    all_cameras = []
-    for home in data_handler.data[CAMERA_DATA_CLASS_NAME].cameras.values():
-        for camera in home.values():
-            all_cameras.append(camera)
-
-    entities = [
-        NetatmoLight(
-            data_handler,
-            camera["id"],
-            camera["type"],
-            camera["home_id"],
+
+    @callback
+    def _create_camera_light_entity(netatmo_device: NetatmoDevice) -> None:
+        if not hasattr(netatmo_device.device, "floodlight"):
+            return
+
+        entity = NetatmoCameraLight(netatmo_device)
+        async_add_entities([entity])
+
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, NETATMO_CREATE_CAMERA_LIGHT, _create_camera_light_entity
         )
-        for camera in all_cameras
-        if camera["type"] == "NOC"
-    ]
+    )
 
-    _LOGGER.debug("Adding camera lights %s", entities)
-    async_add_entities(entities, True)
+    @callback
+    def _create_entity(netatmo_device: NetatmoDevice) -> None:
+        if not hasattr(netatmo_device.device, "brightness"):
+            return
 
+        entity = NetatmoLight(netatmo_device)
+        _LOGGER.debug("Adding light %s", entity)
+        async_add_entities([entity])
 
-class NetatmoLight(NetatmoBase, LightEntity):
+    entry.async_on_unload(
+        async_dispatcher_connect(hass, NETATMO_CREATE_LIGHT, _create_entity)
+    )
+
+
+class NetatmoCameraLight(NetatmoBase, LightEntity):
     """Representation of a Netatmo Presence camera light."""
 
-    _attr_color_mode = ColorMode.ONOFF
     _attr_has_entity_name = True
-    _attr_supported_color_modes = {ColorMode.ONOFF}
 
     def __init__(
         self,
-        data_handler: NetatmoDataHandler,
-        camera_id: str,
-        camera_type: str,
-        home_id: str,
+        netatmo_device: NetatmoDevice,
     ) -> None:
         """Initialize a Netatmo Presence camera light."""
         LightEntity.__init__(self)
-        super().__init__(data_handler)
-
-        self._publishers.append(
-            {"name": CAMERA_DATA_CLASS_NAME, SIGNAL_NAME: CAMERA_DATA_CLASS_NAME}
-        )
-        self._id = camera_id
-        self._home_id = home_id
-        self._model = camera_type
-        self._netatmo_type = TYPE_SECURITY
-        self._device_name: str = self._data.get_camera(camera_id)["name"]
+        super().__init__(netatmo_device.data_handler)
+
+        self._camera = cast(NaModules.NOC, netatmo_device.device)
+        self._id = self._camera.entity_id
+        self._home_id = self._camera.home.entity_id
+        self._device_name = self._camera.name
+        self._model = self._camera.device_type
```

Comment:
These lines of code seem duplicated in each platform. Wouldn't it make more sense to put in the NetatmoBase?

Reply:
Put a generic _device member in the base, and override it's type in the inherited EntityClass for example.

Reply:
While I agree with your suggestions I'd rather not implement this at this stage. This is can not be applied uniformly for all platforms right now.